lib/sql.js

```diff
@@ -156,200 +228,892 @@ SqlConnector.prototype.tableEscaped = function (model) {
  * @param {String} property The property name
  * @returns {String} The escaped column name
  */
-SqlConnector.prototype.columnEscaped = function (model, property) {
+SqlConnector.prototype.columnEscaped = function(model, property) {
   return this.escapeName(this.column(model, property));
 };
 
-function isIdValuePresent(idValue, callback, returningNull) {
+/*!
+ * Check if id value is set
+ * @param idValue
+ * @param cb
+ * @param returningNull
+ * @returns {boolean}
+ */
+function isIdValuePresent(idValue, cb, returningNull) {
   try {
     assert(idValue !== null && idValue !== undefined, 'id value is required');
     return true;
   } catch (err) {
-    process.nextTick(function () {
-      if(callback) callback(returningNull ? null: err);
+    process.nextTick(function() {
+      if (cb) cb(returningNull ? null : err);
     });
     return false;
   }
 }
+
+/**
+ * Convert the id value to the form required by database column
+ * @param {String} model The model name
+ * @param {*} idValue The id property value
+ * @returns {*} The escaped id column value
+ */
+SqlConnector.prototype.idColumnValue = function(model, idValue) {
+  var idProp = this.getDataSource(model).idProperty(model);
+  if (typeof this.toColumnValue === 'function') {
+    return this.toColumnValue(idProp, idValue);
+  } else {
+    return idValue;
+  }
+};
+
+/**
+ * Replace `?` with connector specific placeholders. For example,
+ *
+ * ```
+ * {sql: 'SELECT * FROM CUSTOMER WHERE NAME=?', params: ['John']}
+ * ==>
+ * {sql: 'SELECT * FROM CUSTOMER WHERE NAME=:1', params: ['John']}
+ * ```
+ * @param {ParameterizedSQL|Object} ps Parameterized SQL
+ * @returns {ParameterizedSQL} Parameterized SQL with the connector specific
+ * placeholders
+ */
+SqlConnector.prototype.parameterize = function(ps) {
+  ps = new ParameterizedSQL(ps);
+
+  // The value is parameterized, for example
+  // {sql: 'to_point(?,?)', values: [1, 2]}
+  var parts = ps.sql.split(PLACEHOLDER);
```

Comment:
If I am reading the code correctly, then this implementation is based on an assumption that the PLACEHOLDER will be never a part of a quoted string:

Example:

```
// find all titles that don't contain a question mark
SELECT ? FROM questions WHERE ? NOT LIKE "%?"
==> SELECT :1 FROM question WHERE :2 NOT LIKE "%:3"
       // finds all titles that don't contain string ":3"
```

If this is an acceptable limitation, then I am proposing to describe it clearly in the API docs.